### PR TITLE
client/dcr: Add rpc staking.

### PR DESCRIPTION
**Pull request by JoeGruffins**
part of #2264
~depends on #2290~

Adds a rpc consumer for ticket purchasing.

### CODE REVIEW

File: dex/testing/dcr/create-vspd.sh
[the remark this answers is not in the document]
Comment (JoeGruffins) [reply dex/testing/dcr/create-vspd.sh:1]: Sorry, adding what? An rpc ticket buying loop?

[the remark this answers is not in the document]
Comment (buck54321) [reply dex/testing/dcr/create-vspd.sh:1]: Nevermind. You've got it right. Carry on.

File: client/rpcserver/handlers.go [hunk 1]
```diff
@@ -1584,4 +1654,88 @@ needed to complete a swap.`,
 	stopMarketMakingRoute: {
 		cmdSummary: `Stop market making.`,
 	},
+	stakeStatusRoute: {
+		cmdSummary: `Get stake status. `,
+		returns: `Returns:
+  obj: The staking status.
+    {
+      ticketprice (uint64): The current ticket price in atoms.
```
Comment (buck54321) [inline client/rpcserver/handlers.go:1662]: `ticketprice -> ticketPrice`

File: client/rpcserver/handlers.go [hunk 2]
```diff
@@ -1584,4 +1654,88 @@ needed to complete a swap.`,
...
+      vsp (string): The url of the currently set vsp (voting service provider).
+      isrpc (bool): Whether the wallet is an RPC wallet. False indicates
```
Comment (buck54321) [inline client/rpcserver/handlers.go:1664]: `isrpc -> isRPC`

File: client/rpcserver/handlers.go [hunk 3]
```diff
@@ -1584,4 +1654,88 @@ needed to complete a swap.`,
...
+an spv wallet and allows options that can view and set the vsp.
```
Comment (buck54321) [inline client/rpcserver/handlers.go:1665]: `allows options that can view ...` `->` `enables options to view ...`

File: client/rpcserver/handlers.go [hunk 4]
```diff
@@ -1584,4 +1654,88 @@ needed to complete a swap.`,
...
+            hash (string): The ticket hash as hex.
+            ticketPrice (int): The amount paid for the ticket in atoms.
+            fees (int): The ticket transaction's tx fee.
+            stamp (int): The UNIX time the ticket was purchased.
+            blockheight (int): The block number the ticket was mined.
+          },
+          status: (int) The ticket status. 0 is TicketStatusUnknown. 1 is TicketStatusUnmined.
+2 is TicketStatusImmature. 3 is TicketStatusLive. 4 is TicketStatusVoted.
+5 is TicketStatusMissed. 6 is TicketStatusExpired. 7 is TicketStatusUnspent.
+8 is TicketStatusRevoked.
```
Comment (buck54321) [inline client/rpcserver/handlers.go:1680]:
```suggestion
          status: (int) The ticket status. 0: unknown, 1: unmined, 2: immature, 3: live, 
            4: voted, 5: missed, 6:expired, 7: unspent, 8: revoked.
```

File: client/rpcserver/handlers.go [hunk 5]
```diff
@@ -1584,4 +1654,88 @@ needed to complete a swap.`,
...
+          spender (string): The transaction that votes on or revokes the ticket if available.
+       },
+     ],...
+     stances (obj): Voting policies.
+     {
+       votechoices (array): An array of consensus vote choices.
```
Comment (buck54321) [inline client/rpcserver/handlers.go:1686]: `votechoices -> voteChoices`

File: client/rpcserver/handlers.go [hunk 6]
```diff
@@ -1584,4 +1654,88 @@ needed to complete a swap.`,
...
+       [
+         {
+           agendaid (string): The agenda ID,
+           agendadescription (string): A description of the agenda being voted on.
+           choiceid (string): The current choice.
+           choicedescription (string): A description of the chosen choice.
+         },
+       ],...
+       tspendpolicy (array): An array of TSpend policies.
```
Comment (buck54321) [inline client/rpcserver/handlers.go:1695]: `tspendpolicy -> tSpendPolicy`

File: client/rpcserver/handlers.go [hunk 7]
```diff
@@ -1584,4 +1654,88 @@ needed to complete a swap.`,
...
+       [
+         {
+           hash (string): The TSpend txid.,
+           policy (string): The policy.
+         },
+       ],...
+       treasurypolicy (array): An array of treasury policies.
```
Comment (buck54321) [inline client/rpcserver/handlers.go:1702]: `treasurypolicy -> treasuryPolicy`

File: client/rpcserver/handlers.go [hunk 8]
```diff
@@ -1584,4 +1654,88 @@ needed to complete a swap.`,
...
+		returns: `Returns:
+  array: An array of ticket hashes.`,
+	},
+	setVotingPreferencesRoute: {
+		argsShort:  `(choicesMap) (tSpendPolicyMap) (treasuryPolicyMap)`,
+		cmdSummary: `Cancel an order.`,
+		argsLong: `Args:
+  choicesMap (map[string]string): A map of choices IDs to choice policies.
+  tSpendPolicyMap (map[string]string): A map of tSpend txids to tSpend policies.
+  treasuryPolicyMap (map[string]string): A map of treasury spender public keys to tSpend policies.`,
```
Comment (buck54321) [inline client/rpcserver/handlers.go:1737]:
```suggestion
  choicesMap ({"agendaid": "choiceid", ...}): A map of choices IDs to choice policies.
  tSpendPolicyMap ({"hash": "policy", ...}): A map of tSpend txids to tSpend policies.
  treasuryPolicyMap ({"key": "policy", ...}): A map of treasury spender public keys to tSpend policies.`,
```

**Review by buck54321 — CHANGES_REQUESTED** [review]

File: client/core/core.go [hunk 9]
```diff
@@ -10498,3 +10498,75 @@ func (c *Core) SendShielded(appPW []byte, assetID uint32, toAddr string, amt uin
...
+		return nil, nil, fmt.Errorf("dcr wallet is not a TicketBuyer")
+	}
+	return wallet, ticketBuyer, nil
+}
+
+// StakeStatus returns current staking statuses such as currently owned
+// tickets, ticket price, and current voting preferences. Used for
+// decred ticket purchasing.
+func (c *Core) StakeStatus() (*asset.TicketStakingStatus, error) {
+	_, tb, err := c.dcrStakingWallet()
```
[the remark this answers is not in the document]
Comment (JoeGruffins) [reply client/core/core.go:10522]: I'll make it neutral, no big deal. I also cannot foresee other stakers, and I don't know if the interface would be compatible.

File: client/rpcserver/handlers.go (hunk 1 above)
Comment (JoeGruffins) [reply client/rpcserver/handlers.go:1662]: Ah, fixing these. Was it right to have all of the json tags camel case?

File: client/rpcserver/handlers.go (hunk 3 above)
Comment (JoeGruffins) [reply client/rpcserver/handlers.go:1665]: fixed

File: client/rpcserver/handlers.go (hunk 4 above)
Comment (JoeGruffins) [reply client/rpcserver/handlers.go:1680]: fixed

File: client/rpcserver/handlers.go (hunk 8 above)
Comment (JoeGruffins) [reply client/rpcserver/handlers.go:1737]: fixed

**Comment by JoeGruffins** [comment]
Make core interface asset neutral and cleaning up help messages https://github.com/decred/dcrdex/compare/25a52a4119c86738aeb58f9909da03fc79eefff9..064af34478e9098c95ea04f5dcf4c3cbc70316e0

File: client/rpcserver/handlers.go (hunk 1 above)
Comment (buck54321) [reply client/rpcserver/handlers.go]: Camel-case please.

Comment (JoeGruffins) [reply client/rpcserver/handlers.go]: Leaving as is.

File: client/core/core.go [hunk 10]
```diff
@@ -10498,3 +10498,75 @@ func (c *Core) SendShielded(appPW []byte, assetID uint32, toAddr string, amt uin
...
+// SetVSP sets the VSP provider. Used for ticket purchasing.
+func (c *Core) SetVSP(assetID uint32, addr string) error {
+	_, tb, err := c.stakingWallet(assetID)
+	if err != nil {
+		return err
+	}
+	return tb.SetVSP(addr)
+}
+
+// PurchaseTickets purchases n amout of tickets. Returns the purchased
```
Comment (buck54321) [inline client/core/core.go:10537]: ~~amout~~

File: client/rpcserver/handlers.go [hunk 11]
```diff
@@ -1584,4 +1658,86 @@ needed to complete a swap.`,
 	stopMarketMakingRoute: {
 		cmdSummary: `Stop market making.`,
 	},
+	stakeStatusRoute: {
+		cmdSummary: `Get stake status. `,
+		returns: `Returns:
```
Comment (buck54321) [inline client/rpcserver/handlers.go:1663]: + `assetID (int)` arg

File: client/rpcserver/handlers.go [hunk 12]
```diff
@@ -1584,4 +1658,86 @@ needed to complete a swap.`,
...
+         },
+       ],...
+     }
+  }`,
+	},
+	setVSPRoute: {
+		argsShort:  `"addr"`,
+		cmdSummary: `Set a vsp by url.`,
+		argsLong: `Args:
+  addr (string): The vsp's url.`,
```
Comment (buck54321) [inline client/rpcserver/handlers.go:1718]: + `assetID (int)` arg.

File: client/rpcserver/handlers.go [hunk 13]
```diff
@@ -1584,4 +1658,86 @@ needed to complete a swap.`,
...
+		returns: `Returns:
+  array: An array of ticket hashes.`,
+	},
+	setVotingPreferencesRoute: {
+		argsShort:  `(choicesMap) (tSpendPolicyMap) (treasuryPolicyMap)`,
+		cmdSummary: `Cancel an order.`,
+		argsLong: `Args:
+  choicesMap ({"agendaid": "choiceid", ...}): A map of choices IDs to choice policies.
+  tSpendPolicyMap ({"hash": "policy", ...}): A map of tSpend txids to tSpend policies.
+  treasuryPolicyMap ({"key": "policy", ...}): A map of treasury spender public keys to tSpend policies.`,
```
Comment (buck54321) [inline client/rpcserver/handlers.go:1739]: + `assetID (int)` arg

**Review by buck54321 — CHANGES_REQUESTED** [review]

File: client/rpcserver/handlers.go [hunk 14]
```diff
@@ -1584,4 +1658,86 @@ needed to complete a swap.`,
...
+  string: The message "` + fmt.Sprintf(setVSPStr, "[vsp url]") + `"`,
+	},
+	purchaseTicketsRoute: {
+		pwArgsShort: `"appPass"`,
+		argsShort:   `num`,
+		cmdSummary:  `Purchase some tickets.`,
+		pwArgsLong: `Password Args:
+  appPass (string): The DEX client password.`,
+		argsLong: `Args:
+  num (int): The number of tickets to purchase`,
```
[the remark this answers is not in the document]
Comment (JoeGruffins) [reply client/rpcserver/handlers.go:1729]: Ah shit. I was thinking about this the whole time and then forgot at the end.

[the remark this answers is not in the document]
Comment (JoeGruffins) [reply client/rpcserver/handlers.go]: added

File: client/rpcserver/handlers.go (hunk 11 above)
Comment (JoeGruffins) [reply client/rpcserver/handlers.go]: added

File: client/rpcserver/handlers.go (hunk 12 above)
Comment (JoeGruffins) [reply client/rpcserver/handlers.go]: added

File: client/rpcserver/handlers.go (hunk 13 above)
Comment (JoeGruffins) [reply client/rpcserver/handlers.go]: added

File: client/core/core.go (hunk 10 above)
Comment (JoeGruffins) [reply client/core/core.go]: removed

**Comment by JoeGruffins** [comment]
Added `assetID` arguments to rpc help messages. https://github.com/decred/dcrdex/compare/f4f9f021b11a9ad7e568fba86d68a5f66781179e..a465c48681aca261a21c62f62bf24575c9820227

File: client/asset/dcr/rpcwallet.go [hunk 15]
```diff
@@ -912,26 +912,36 @@ func (w *rpcWallet) Tickets(ctx context.Context) ([]*asset.Ticket, error) {
 			w.log.Errorf("GetTransaction error for ticket %s: %v", h, err)
 			continue
 		}
+		blkHash, err := chainhash.NewHashFromStr(tx.BlockHash)
```
Comment (itswisdomagain) [inline client/asset/dcr/rpcwallet.go:915]: How would this work for unmined tickets? Or are all tickets guaranteed to be included in a block?

File: client/core/core.go [hunk 16]
```diff
@@ -10498,3 +10498,75 @@ func (c *Core) SendShielded(appPW []byte, assetID uint32, toAddr string, amt uin
...
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Send tickets bought notification.
+	//subject, details := c.formatDetails(TopicSendSuccess, sentValue, unbip(assetID), address, coin)
+	//c.notify(newSendNote(TopicSendSuccess, subject, details, db.Success))
+
+	c.updateAssetBalance(assetID)
+	return tb.PurchaseTickets(n)
```
Comment (itswisdomagain) [inline client/core/core.go:10559]: Should the balance updating happen _after_ the ticket purchase, if successful? Same with the ntfn TODO a few lines above.

File: client/asset/dcr/rpcwallet.go (hunk 15 above)
Comment (JoeGruffins) [reply client/asset/dcr/rpcwallet.go:915]: I believe it errors and I saw this while testing with a block hash of all zeros....

File: client/core/core.go (hunk 16 above)
Comment (JoeGruffins) [reply client/core/core.go:10559]: Oh yeah, after does make more sense, hmm

File: client/asset/dcr/rpcwallet.go (hunk 15 above)
Comment (JoeGruffins) [reply client/asset/dcr/rpcwallet.go:915]: Testing out on simnet, it seems vsp is only including mined tickets in the results. @buck54321 do we need confirming tickets for native wallets? Or we can ignore these here as well so they act the same. Probably we want the unconfirmed tickets in results...

Comment (buck54321) [reply client/asset/dcr/rpcwallet.go:915]: I saw that too, but I figured we could follow up since it's not particular to RPC staking.

Comment (JoeGruffins) [reply client/asset/dcr/rpcwallet.go:915]: Will set unmined block heights to -1 here and leave a TODO in the vsp wallet logic.

File: client/core/core.go (hunk 16 above)
Comment (JoeGruffins) [reply client/core/core.go]: fixed

**Comment by JoeGruffins** [comment]
isSPV -> isNative as pointed out by @itswisdomagain an rpcwallet can also be SPV https://github.com/decred/dcrdex/compare/a465c48681aca261a21c62f62bf24575c9820227..ea84ec5eac9a56a29524596167db2d165a21b0c5

**Pull request merged**